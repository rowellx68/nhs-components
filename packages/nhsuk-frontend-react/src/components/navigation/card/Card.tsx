'use client';

import React from 'react';
import { Factory, factory } from '@/internal/factory/factory';
import { AsElementProps, ElementProps } from '@/types/shared';
import { CardProvider, useCardContext } from './Card.context';
import clsx from 'clsx';
import {
  PolymorphicFactory,
  polymorphicFactory,
} from '@/internal/factory/polymorphic-factory';
import { Base } from '@/components/core/base/Base';
import {
  Heading,
  HeadingProps,
} from '@/components/styles/typography/heading/Heading';
import { ChevronRightCircleIcon } from '@/icons/chevron-right-circle/ChevronRightCircle';
import {
  Column,
  ColumnProps,
  Row,
  RowProps,
} from '@/components/styles/layout/grid/Grid';
import { VisuallyHidden } from '@/components/core/visually-hidden/VisuallyHidden';

export type CardVariant =
  | 'primary'
  | 'secondary'
  | 'feature'
  | 'non-urgent'
  | 'urgent'
  | 'emergency';

export type CardProps = (
  | {
      clickable?: boolean;
      variant?: CardVariant;
    }
  | {
      clickable?: false;
      variant: Extract<
        CardVariant,
        'feature' | 'non-urgent' | 'urgent' | 'emergency'
      >;
    }
) &
  ElementProps<'div'>;

const careCardVariants = ['non-urgent', 'urgent', 'emergency'];

type CardFactory = Factory<{
  props: CardProps;
  ref: HTMLDivElement;
  staticComponents: {
    Image: typeof CardImage;
    Content: typeof CardContent;
    Heading: typeof CardHeading;
    Description: typeof CardDescription;
    Link: typeof CardLink;
    Group: typeof CardGroup;
    GroupItem: typeof CardGroupItem;
  };
}>;

const Card = factory<CardFactory>(
  ({ variant, clickable, className, ...props }, ref) => {
    return (
      <CardProvider value={{ variant: variant }}>
        <div
          className={clsx(
            'nhsuk-card',
            {
              'nhsuk-card--clickable': clickable,
              [`nhsuk-card--${variant}`]:
                variant && !careCardVariants.includes(variant),
              [`nhsuk-card--care nhsuk-card--care--${variant}`]:
                variant && careCardVariants.includes(variant),
            },
            className,
          )}
          {...props}
          ref={ref}
        />
      </CardProvider>
    );
  },
);

export type CardImageProps = ElementProps<'img'>;

type CardImageFactory = PolymorphicFactory<{
  props: CardImageProps;
  defaultComponent: 'img';
  defaultRef: HTMLImageElement;
}>;

const CardImage = polymorphicFactory<CardImageFactory>(
  (
    {
      className,
      as: component = 'img',
      ...props
    }: CardImageProps & AsElementProps,
    ref,
  ) => {
    return (
      <Base
        as={component}
        className={clsx('nhsuk-card__img', className)}
        ref={ref}
        {...props}
      />
    );
  },
);

export type CardContentProps = ElementProps<'div'>;

const CardContent = ({ className, children, ...props }: CardContentProps) => {
  const { variant } = useCardContext();
  return (
    <div
      className={clsx(
        'nhsuk-card__content',
        { [`nhsuk-card__content--${variant}`]: variant },
        className,
      )}
      {...props}
    >
      {children}
      {variant === 'primary' && <ChevronRightCircleIcon />}
    </div>
  );
};

export type CardHeadingProps = {
  /**
   * This will only be used if the Card variant is either 'non-urgent', 'urgent' or 'emergency'.
   */
  visuallyHiddenText?: string;
} & HeadingProps;

const CardHeading = ({
  className,
  children,
  as: component = 'h2',
  visuallyHiddenText,
  ...props
}: CardHeadingProps) => {
  const { variant } = useCardContext();
  const careCard = variant && careCardVariants.includes(variant);

  const wrapperProps = {
    as: careCard ? 'div' : React.Fragment,
    ...(careCard ? { className: 'nhsuk-card--care__heading-container' } : {}),
  };

  const headingTextProps = {
    as: careCard ? 'span' : React.Fragment,
    ...(careCard ? { role: 'text' } : {}),
  };

  return (
    <Base<any> {...wrapperProps}>
      <Heading
        as={component}
        className={clsx(
          {
            'nhsuk-card__heading':
              !variant || !careCardVariants.includes(variant),
            'nhsuk-card--care__heading': careCard,
            'nhsuk-card__heading--feature nhsuk-u-font-size-24':
              variant === 'feature',
          },
          className,
        )}
        {...props}
      >
        <Base<any> {...headingTextProps}>
          {careCard && <VisuallyHidden>{visuallyHiddenText}</VisuallyHidden>}
          {children}
        </Base>
      </Heading>
      {careCard && (
        <span className="nhsuk-card--care__arrow" aria-hidden="true" />
      )}
    </Base>
  );
};

export type CardDescriptionProps = ElementProps<'p'>;

const CardDescription = ({ className, ...props }: CardDescriptionProps) => {
  return (
    <p className={clsx('nhsuk-card__description', className)} {...props} />
  );
};

export type CardLinkProps = ElementProps<'a'>;

type CardLinkFactory = PolymorphicFactory<{
  props: CardLinkProps;
  defaultComponent: 'a';
  defaultRef: HTMLAnchorElement;
}>;

const CardLink = polymorphicFactory<CardLinkFactory>(
  (
    {
      className,
      as: component = 'a',
      ...props
    }: CardLinkProps & AsElementProps,
    ref,
  ) => {
    return (
      <Base
        as={component}
        className={clsx('nhsuk-card__link', className)}
        ref={ref}
        {...props}
      />
    );
  },
);

export type CardGroupProps = RowProps;

const CardGroup = ({ className, ...props }: CardGroupProps) => {
  return <Row className={clsx('nhsuk-card-group', className)} {...props} />;
};

export type CardGroupItemProps = ColumnProps;

const CardGroupItem = ({ className, ...props }: CardGroupItemProps) => {
  return (
    <Column className={clsx('nhsuk-card-group__item', className)} {...props} />
  );
};

Card.displayName = 'Card';
CardImage.displayName = 'Card.Image';
CardContent.displayName = 'Card.Content';
CardHeading.displayName = 'Card.Heading';
CardDescription.displayName = 'Card.Description';
CardLink.displayName = 'Card.Link';
CardGroup.displayName = 'Card.Group';
CardGroupItem.displayName = 'Card.GroupItem';

Card.Image = CardImage;
Card.Content = CardContent;
Card.Heading = CardHeading;
Card.Description = CardDescription;
Card.Link = CardLink;
Card.Group = CardGroup;
Card.GroupItem = CardGroupItem;

export {
  Card,
  CardImage,
  CardContent,
  CardHeading,
  CardDescription,
  CardLink,
  CardGroup,
  CardGroupItem,
};
