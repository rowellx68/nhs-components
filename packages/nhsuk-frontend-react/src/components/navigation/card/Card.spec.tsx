import React from 'react';
import { it, expect } from 'vitest';
import { render } from '@testing-library/react';
import { Card } from './Card';

it('should render the card component', () => {
  const { container } = render(
    <Card.Group>
      <Card.GroupItem width="one-third">
        <Card clickable>
          <Card.Content>
            <Card.Heading as="h5" size="xs">
              <Card.Link href="#">Order a repeat prescription</Card.Link>
            </Card.Heading>
          </Card.Content>
        </Card>
      </Card.GroupItem>
    </Card.Group>,
  );

  expect(container).toMatchSnapshot();
  expect(container.querySelector('h5')).toBeInTheDocument();
});

it('should render the card component with a primary variant', () => {
  const { container } = render(
    <Card.Group>
      <Card.GroupItem width="two-thirds">
        <Card clickable variant="primary">
          <Card.Content>
            <Card.Heading as="h2" size="m">
              <Card.Link href="#">Introduction to care and support</Card.Link>
            </Card.Heading>
            <Card.Description>
              A quick guide for people who have care and support needs and their
              carers
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
    </Card.Group>,
  );

  expect(container).toMatchSnapshot();
  expect(container.querySelector('h2')).toBeInTheDocument();
  expect(
    container.querySelector('.nhsuk-icon__chevron-right-circle'),
  ).toBeInTheDocument();
});

it('should render the card component with a secondary variant', () => {
  const { container } = render(
    <Card.Group>
      <Card.GroupItem width="two-thirds">
        <Card clickable variant="secondary">
          <Card.Content>
            <Card.Heading as="h2" size="m">
              <Card.Link href="#">Urgent and emergency care services</Card.Link>
            </Card.Heading>
            <Card.Description>
              Services the NHS provides if you need urgent or emergency medical
              help
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
    </Card.Group>,
  );

  expect(container).toMatchSnapshot();
  expect(container.querySelector('h2')).toBeInTheDocument();
});

it('should render the card component with an image', () => {
  const { container } = render(
    <Card.Group>
      <Card.GroupItem width="one-third">
        <Card clickable>
          <Card.Image
            src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
            alt="A person wearing a face mask"
          />
          <Card.Content>
            <Card.Heading as="h2" size="m">
              <Card.Link href="#">Exercise</Card.Link>
            </Card.Heading>

            <Card.Description>
              Programmes, workouts and tips to get you moving and improve your
              fitness and wellbeing
            </Card.Description>
          </Card.Content>
        </Card>
      </Card.GroupItem>
    </Card.Group>,
  );

  expect(container).toMatchSnapshot();
  expect(container.querySelector('h2')).toBeInTheDocument();
  expect(container.querySelector('img')).toBeInTheDocument();
});
