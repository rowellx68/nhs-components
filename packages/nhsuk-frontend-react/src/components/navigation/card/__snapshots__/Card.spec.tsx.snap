// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should render the card component 1`] = `
<div>
  <div
    class="nhsuk-grid-row nhsuk-card-group"
  >
    <div
      class="nhsuk-grid-column-one-third nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card nhsuk-card--clickable"
      >
        <div
          class="nhsuk-card__content"
        >
          <h5
            class="nhsuk-heading-xs nhsuk-card__heading"
          >
            <a
              class="nhsuk-card__link"
              href="#"
            >
              Order a repeat prescription
            </a>
          </h5>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the card component with a primary variant 1`] = `
<div>
  <div
    class="nhsuk-grid-row nhsuk-card-group"
  >
    <div
      class="nhsuk-grid-column-two-thirds nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card nhsuk-card--clickable nhsuk-card--primary"
      >
        <div
          class="nhsuk-card__content nhsuk-card__content--primary"
        >
          <h2
            class="nhsuk-heading-m nhsuk-card__heading"
          >
            <a
              class="nhsuk-card__link"
              href="#"
            >
              Introduction to care and support
            </a>
          </h2>
          <p
            class="nhsuk-card__description"
          >
            A quick guide for people who have care and support needs and their carers
          </p>
          <svg
            aria-hidden="true"
            class="nhsuk-icon nhsuk-icon__chevron-right-circle"
            focusable="false"
            height="27"
            width="27"
            xmlns="http://www.w3.org/2000/svg"
          >
            <circle
              cx="13.333"
              cy="13.333"
              fill=""
              r="13.333"
            />
            <g
              data-name="Group 1"
              fill="none"
              stroke="#fff"
              stroke-linecap="round"
              stroke-miterlimit="10"
              stroke-width="2.667"
            >
              <path
                d="M15.438 13l-3.771 3.771"
              />
              <path
                d="M11.667 9.229L15.438 13"
                data-name="Path"
              />
            </g>
          </svg>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the card component with a secondary variant 1`] = `
<div>
  <div
    class="nhsuk-grid-row nhsuk-card-group"
  >
    <div
      class="nhsuk-grid-column-two-thirds nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card nhsuk-card--clickable nhsuk-card--secondary"
      >
        <div
          class="nhsuk-card__content nhsuk-card__content--secondary"
        >
          <h2
            class="nhsuk-heading-m nhsuk-card__heading"
          >
            <a
              class="nhsuk-card__link"
              href="#"
            >
              Urgent and emergency care services
            </a>
          </h2>
          <p
            class="nhsuk-card__description"
          >
            Services the NHS provides if you need urgent or emergency medical help
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the card component with an image 1`] = `
<div>
  <div
    class="nhsuk-grid-row nhsuk-card-group"
  >
    <div
      class="nhsuk-grid-column-one-third nhsuk-card-group__item"
    >
      <div
        class="nhsuk-card nhsuk-card--clickable"
      >
        <img
          alt="A person wearing a face mask"
          class="nhsuk-card__img"
          src="https://assets.nhs.uk/prod/images/A_0218_exercise-main_FKW1X7.width-690.jpg"
        />
        <div
          class="nhsuk-card__content"
        >
          <h2
            class="nhsuk-heading-m nhsuk-card__heading"
          >
            <a
              class="nhsuk-card__link"
              href="#"
            >
              Exercise
            </a>
          </h2>
          <p
            class="nhsuk-card__description"
          >
            Programmes, workouts and tips to get you moving and improve your fitness and wellbeing
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
